# Genome-resolved metagenomics and functional analysis. 

A subset of the samples was used for shot-gun sequencing (150PE) with one replicate per time point. Processing of data was adopted from the Rasmussen et al. 2021 [62]. In brief, reads were quality controlled, using FastQC (v0.11.8). Removal of adapters and low-quality reads was done with AdapterRemoval (v2.3.3) [63], with a quality base of 33 (default) and a minimum length of 50 bp. Duplicates were removed, and reads were re-paired to remove singletons, using bbmap (v.38.35). To increase assembly efficiency by reducing eukaryotic contaminants, data were filtered for the bryozoan (GCA_914767715.1) genome and human (HG19) genome using minimap (V?) [64]. Filtered data were co-assembled using MegaHIT (v1.2.9) with a minimal length of 1000 bp per scaffold, using the meta-large flag for large and complex metagenomes. We used the anviâ€™o pipeline [65] for binning and refinement of genome-resolved metagenomics as described in Rasmussen et al. 2021 [62]. Anviâ€™o was used to profile the scaffolds and identify ORF using Prodigal (v2.6.3) [66] and HMMER (v.3.355) matching archaeal, bacterial [67], protista (http://merenlab.org/delmont-euk-scgs) single-copy core gene (SCG) collections and the universal marker for RNA polymerase A and B [68]. Completeness and redundancy of MAGs were calculated based on SCGs in anviâ€™o databases. MAGs with completeness > 50 % and < 10 % redundancy were further processed. Predicted gene functions were annotated using Pfam [69] and COG [70]. BGCs and functional annotation were predicted by antiSMASH (v6.0) [71] and BiGSCAPE [72]. Differential abundance analysis of all genes recovered from the co-assembly was carried out using DESeq2 (v1.36.0)[73] in R (v4.2.2).
